src/_P065_DRF0299_MP3.ino

```diff
@@ -47,179 +53,214 @@ boolean Plugin_065(byte function, struct EventStruct *event, String& string)
+        PCONFIG(0) = value;
+        Plugin_065_SetVol(value);
+        success = true;
+      }
 
-          int8_t vol = param.toInt();
-          if (vol == 0) vol = 30;
-          PCONFIG(0) = vol;
-          Plugin_065_SetVol(vol);
-          log += vol;
+      if (valueValid && (command == F("eq")))
+      {
+        Plugin_065_SetEQ(value);
+        success = true;
+      }
 
-          addLog(LOG_LEVEL_INFO, log);
-          success = true;
-        }
+      if (valueValid && (command == F("mode")))
+      {
+        Plugin_065_SetMode(value);
+        success = true;
+      }
 
-        if (command == F("eq"))
-        {
-          String log = F("MP3  : eq=");
+      if (valueValid && (command == F("repeat")))
+      {
+        Plugin_065_SetRepeat(value);
+        success = true;
+      }
 
-          int8_t eq = param.toInt();
-          Plugin_065_SetEQ(eq);
-          log += eq;
+      if (success && loglevelActiveFor(LOG_LEVEL_INFO)) {
+        String log;
+        log.reserve(48);
+        log  = F("MP3  : ");
+        log += command;
 
-          addLog(LOG_LEVEL_INFO, log);
-          success = true;
+        if (command != F("stop")) {
+          log += '=';
+          log += value;
         }
-
-        break;
+        addLog(LOG_LEVEL_INFO, log);
       }
+      break;
+    }
   }
   return success;
 }
 
-
 void Plugin_065_Play(uint16_t track)
 {
+  if (track < 0) { track = 0; }
+
+  if (track > 2999) { track = 2999; }
+
   Plugin_065_SendCmd(0x03, track);
 }
 
 void Plugin_065_SetVol(int8_t vol)
 {
-  if (vol < 1) vol = 1;
-  if (vol > 30) vol = 30;
+  if (vol < 1) { vol = 1; }
+
+  if (vol > 30) { vol = 30; }
+
   Plugin_065_SendCmd(0x06, vol);
 }
 
 void Plugin_065_SetEQ(int8_t eq)
 {
-  if (eq < 0) eq = 0;
-  if (eq > 5) eq = 5;
+  if (eq < 0) { eq = 0; }
+
+  if (eq > 5) { eq = 5; }
+
   Plugin_065_SendCmd(0x07, eq);
 }
 
+void Plugin_065_SetMode(int8_t mode)
+{
+  if (mode < 0) { mode = 0; }
+
+  if (mode > 3) { mode = 3; }
+
+  Plugin_065_SendCmd(0x08, mode);
+}
+
+void Plugin_065_SetRepeat(int8_t repeat)
+{
+  if (repeat < 0) { repeat = 0; }
+
+  if (repeat > 1) { repeat = 1; }
+
+  Plugin_065_SendCmd(0x11, repeat);
+}
+
 void Plugin_065_SendCmd(byte cmd, int16_t data)
 {
-  if (!P065_easySerial)
+  if (!P065_easySerial) {
     return;
+  }
 
   byte buffer[10] = { 0x7E, 0xFF, 0x06, 0, 0x00, 0, 0, 0, 0, 0xEF };
 
   buffer[3] = cmd;
   buffer[5] = data >> 8;   // high byte
-  buffer[6] = data & 0xFF;   // low byte
+  buffer[6] = data & 0xFF; // low byte
 
   int16_t checksum = -(buffer[1] + buffer[2] + buffer[3] + buffer[4] + buffer[5] + buffer[6]);
-  buffer[7] = checksum >> 8;   // high byte
-  buffer[8] = checksum & 0xFF;   // low byte
 
-  P065_easySerial->write(buffer, 10);   //Send the byte array
+  buffer[7] = checksum >> 8;          // high byte
+  buffer[8] = checksum & 0xFF;        // low byte
+
+  P065_easySerial->write(buffer, 10); // Send the byte array
 
   String log = F("MP3  : Send Cmd ");
```

Comment:
Shouldn't this line also be in the if(loglevelActiveFor...) check?

Reply:
missed that